Pre-identity policy

rlenglet opened the pull request:
Define `reserved:init` label and set it on endpoints with no labels.

Always enable policy for `reserved:init` endpoints in `default` mode.

Add `init` as supported entity in the API (`FromEntity` / `ToEntity`).

Allocate reserved identities for reserved labels, esp. allocate ID `4` to health daemon endpoints (`health` entities), and allocate ID `5` to initializing endpoints (`init` entities). Allocate such entity IDs synchronously, since that doesn't require any kvstore interactions.

Simplify and clean up `cilium-docker`.
Remove the constraint on endpoint state after creation, since after creation endpoints can now be in different states: `creating`, `waiting-for-identity`, `not-ready`, etc.
Create the veth pair immediately on endpoint creation, so that the LXC and node MAC addresses are known at endpoint creation time, like with the CNI plugin.
Remove the now-unnecessary `PATCH /endpoint/{id}` API call to Cilium agent, since all the information that used to be passed in the `PATCH` call is now passed in the `PUT` call. The remaining state change triggered by the `PATCH` call was also unnecessary. This reduces the number of regenerations required for each endpoint.

Fix the endpoint state machine to support changing endpoint's labels.

Fix the identity label update APIs.
Make `PUT /endpoint/{id}` return status code 400 in case of failure to set labels, instead of code 404 which was not specified in the API. Validate the labels passed by the client.
`PUT /endpoint/{id}` now sets orchestration labels, not custom labels, to be consistent with `PATCH /endpoint/{id}`.

Fix Runtime Policies tests to handle initializing endpoints with the `reserved:init` identity, and to test policies applying to `reserved:init`.

Fixes: https://github.com/cilium/cilium/issues/3855
Fixes: https://github.com/cilium/cilium/issues/3895
Signed-off-by: Romain Lenglet <romain@covalent.io>

```release-note
A new `reserved:init` label is given to every new endpoint that has no label, e.g. an endpoint that is still initializing. Policy rules can be applied to the endpoints in that state using the new `init` entity.
```

rlenglet commented on pkg/endpoint/policy.go:
```diff
@@ -693,22 +693,13 @@ func (e *Endpoint) updateNetworkPolicy(owner Owner) error {
-		// the policy. Bpf generation calls us again after PolicyMap is created.
-		// In dry mode we are called with a nil PolicyMap.
-
+	if e.Consumable == nil {
```
Note that I had to modify this to make the tests pass when setting labels on endpoint creation.
That codepath was dead code and was broken.

ianvernon commented on daemon/endpoint_test.go:
```diff
@@ -62,17 +63,26 @@ func (ds *DaemonSuite) TestEndpointAdd(c *C) {
+
+	// Check that the policy is always enforced for endpoints with the
+	// reserved:init label, regardless of the mode that is configured.
+	policy.SetPolicyEnabled(endpoint.NeverEnforce)
```
This shouldn't be the case in my opinion. If policy enforcement for the daemon is set to `NeverEnforce`, then the policy corresponding to `reserved:init` should not apply at all. In other words, the policy applying to `reserved:init` should behave just like any other policy in conduit with the daemon's policy enforcement mode.

rlenglet replied:
Fixed.

ianvernon commented on daemon/policy.go:
```diff
@@ -54,7 +54,13 @@ func (d *Daemon) TriggerPolicyUpdates(force bool) *sync.WaitGroup {
 func (d *Daemon) EnableEndpointPolicyEnforcement(e *endpoint.Endpoint) (ingress bool, egress bool) {
-	// First check if policy enforcement should be enabled at the daemon level.
+	// If the endpoint has the reserved:init label, i.e. if it has not yet
+	// received any identity, always enforce policy (default deny).
```
Same as my comment above:

This shouldn't be the case in my opinion. If policy enforcement for the daemon is set to `NeverEnforce`, then the policy corresponding to `reserved:init` should not apply at all. In other words, the policy applying to `reserved:init` should behave just like any other policy in conduit with the daemon's policy enforcement mode.

rlenglet replied:
Fixed.

jrajahalme commented on daemon/endpoint.go:
```diff
@@ -160,21 +160,25 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
+			logfields.IdentityLabels: logfields.Repr(addLabels),
+		}).WithError(err).Error("Could not add labels while creating an ep")
+		return code, err
+	}
```
I recall adding the 'XXX' comment earlier, and still wonder why we are not deleting the just created endpoint if ceateEndpoint() fails?

rlenglet replied:
I don't have any opinion on that.

ianvernon replied:
Nit: while creating an ep --> while creating endpoint .

tgraf replied:
The labels should really be updated before it is being added. Fortunately this could only fail in theory as ModifyIdentityLabels can only fail if labels are being deleted. I'm fixing this in PR https://github.com/cilium/cilium/pull/4065

tgraf replied:
Just saw that you fixed this in a later commit

jrajahalme commented on daemon/endpoint.go:
```diff
@@ -223,7 +227,7 @@ func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Respon
 
 	// Validate the template. Assignment afterwards is atomic.
 	// Note: newEp's labels are ignored.
```
Any idea what these comments mean and whether they still apply?

rlenglet replied:
Yes, they still apply. The labels are indeed ignored in `PATCH /endpoint/{id}`.

jrajahalme commented on daemon/endpoint.go:
```diff
@@ -312,8 +316,10 @@ func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Respon
+	// known and hasn't changed, bump it to ready state immediately to force
+	// re-generation.
+	if ep.GetStateLocked() == endpoint.StateWaitingForIdentity &&
+		ep.SecurityIdentity != nil && ep.SecurityIdentity.Labels.Equals(addLabels) {
```
The old comment was a bit misleading. What we did here was to move the endpoint out of the waiting-for-identity state if the endpoint (now) has an identity. I guess the intent here is to keep the endpoint in the waiting-for-identity state even if it has the init label? Is this the case with the additional condition here?

To inaccuracy with the old comment is that all changes (as marked by the 'changed' flag) currently trigger regeneration also in the waiting-for-identity state.

Not directly related to this, but I wonder what happens to changes made in a PATCH that happen when an endpoint is in "regenerating" state? We fail to trigger/queue a new build?

rlenglet replied:
The fix here is to handle identity changes.

I don't know what happens in all the cases with `PATCH /endpoint/{id}`. Only a few cases were actually covered by tests. And now that I remove calls to `PATCH /endpoint/{id}` from the `cilium-docker` plugin, I suspect this is dead code now anyway.

aanm replied:
I don't get this, shouldn't this be `&& !ep.SecurityIdentity.Labels.Equals(addLabels)`?

rlenglet replied:
I removed this chunk. It was not necessary anymore.

jrajahalme commented on daemon/policy_test.go:
```diff
@@ -444,12 +444,15 @@ func (ds *DaemonSuite) TestRemovePolicy(c *C) {
 	}()
+	buildSuccess := <-e.Regenerate(ds.d, "test")
+	c.Assert(buildSuccess, Equals, true)
+
```
Why did we need to add an additional regeneration step here?

rlenglet replied:
IIRC, that's because we now have to go through 2 generations to have the policy generated for the endpoint's final identity. The 1st generation is for the `reserved:init` identity, the 2nd is for the final identity.

tgraf replied:
Please add a comment to indicate this

rlenglet replied:
I removed this. It is not needed anymore.

jrajahalme commented on pkg/endpoint/endpoint.go:
```diff
@@ -1933,7 +1933,7 @@ func (e *Endpoint) SetStateLocked(toState, reason string) bool {
 	case StateReady:
 		switch toState {
-		case StateDisconnecting, StateWaitingToRegenerate:
+		case StateWaitingForIdentity, StateDisconnecting, StateWaitingToRegenerate:
```
IMO this is wrong. Endpoint can not transition from ready to waiting-for-identity state. Can you give an example where this transition is valid?

rlenglet replied:
We discussed this in person. This is necessary to handle endpoints changing their identity (e.g. have new labels). That means that changing identity was broken.

jrajahalme commented on pkg/endpoint/endpoint.go:
```diff
@@ -1946,7 +1946,7 @@ func (e *Endpoint) SetStateLocked(toState, reason string) bool {
 		switch toState {
 		// Note that transitions to waiting-to-regenerate state
-		case StateDisconnecting:
+		case StateWaitingForIdentity, StateDisconnecting:
```
So far the design has been that an endpoint remains in the waiting-for-identity state also during the initial generation of endpoint's bpf state & program. In fact, an endpoint should transition away from the waiting-for-identity state only after it's identity is resolved.

IMO the addition of the init labels should not change the state machine at all, but an endpoint with the init label should be considered as one without an identity as far as the endpoint's state is considered.

rlenglet replied:
Discussed this in person. All the changes to the state machine are required to support changing an endpoint identity. Moving from `reserved:init` to another identity based on the endpoint's labels is only one case of identity change.

jrajahalme commented on pkg/endpoint/endpoint.go:
```diff
@@ -1955,8 +1955,8 @@ func (e *Endpoint) SetStateLocked(toState, reason string) bool {
-		// waiting-to-regenerate state.
-		case StateDisconnecting, StateWaitingToRegenerate:
+		// waiting-for-identity or waiting-to-regenerate state.
+		case StateWaitingForIdentity, StateDisconnecting, StateWaitingToRegenerate:
```
Same here.

jrajahalme commented on pkg/endpoint/policy.go:
```diff
@@ -601,6 +601,12 @@ func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (
+		e.getLogger().Warn("Endpoint lacks identity, skipping policy calculation")
+		return false, nil
+	}
+
```
This seems to prevent nil dereference for the Lock() below. If this is the case, should this change be backported separately? Or was this needed due to other changes in this PR, like for example that while the endpoint has the reserved:init label it does not have a consumable?

rlenglet replied:
This is relevant only for unit tests, in the new unit tests I added.
So this doesn't need to be backported.

joestringer replied:
Do we still need the additional check for `c.ID == 0` below, with this change? Is it ever valid to set a consumable with id=0?

aanm replied:
@rlenglet `owner.DryModeEnabled()` set on line 522 should prevent this code from being reached in unit tests

rlenglet replied:
It was necessary to make unit tests pass.
However, this chunk is no longer relevant, as we got rid of Consumable altogether.

jrajahalme commented on pkg/identity/allocator.go:
```diff
@@ -91,6 +91,21 @@ func InitIdentityAllocator(owner IdentityAllocatorOwner) {
+func IdentityAllocationIsLocal(lbls labels.Labels) bool {
+	// If there only one label with the "reserved" source and a well-known key,
+	// use the well-known identity for that key.
+	if reservedIdentity := LookupReservedIdentity(lbls); reservedIdentity != nil {
```
This comment is the same as elsewhere with LookupReservedIdentity() call, even though context here is different. Suggest updating the comment here to match the context.

jrajahalme replied:
Could also just `return LookupReservedIdentity(lbls) != nil`

rlenglet replied:
Fixed the comment.
Returned directly.

jrajahalme commented on pkg/identity/cache.go:
```diff
@@ -97,6 +93,25 @@ func LookupIdentity(lbls labels.Labels) *Identity {
+		return nil
+	}
+	for _, lbl := range lbls {
+		if lbl.Source != labels.LabelSourceReserved {
```
Since we only ever return non-nil if there is (one) label, it would be clearer to index that label directly without the `for .. range` statement.

rlenglet replied:
Done.

jrajahalme commented on pkg/labels/labels.go:
```diff
@@ -245,6 +249,24 @@ func (l Labels) AppendPrefixInKey(prefix string) Labels {
+	for k, lbl1 := range l {
+		if lbl2, ok := other[k]; ok {
+			if lbl1.Source == lbl2.Source && lbl1.Key == lbl2.Key && lbl1.Value == lbl2.Value {
+				continue ThisLabelsLoop
```
There is only one loop, so the label for the loop is not needed?

jrajahalme replied:
The fact that you used `ThisLabelsLoop` made me look for the other loop, which does not exist. Looping over `l` is sufficient because a set of labels can never contain duplicates, and we always take care to not violate that invariant, right?

rlenglet replied:
There was another loop in the first iterations. :)
Fixed.

jrajahalme commented on plugins/cilium-docker/driver/driver.go:
```diff
@@ -417,28 +430,11 @@ func (driver *driver) joinEndpoint(w http.ResponseWriter, r *http.Request) {
-	if old.Status.State != models.EndpointStateCreating {
-		sendError(w, "Error: endpoint not in creating state", http.StatusBadRequest)
-		return
-	}
```
Is this change needed if this PR would not change the observable endpoint state?

rlenglet replied:
Correct. But that was brittle anyway.

rlenglet replied:
For the record, I had to remove this check, because now an endpoint can be in any state shortly after creation: waiting-for-identity, waiting-to-regenerate, regenerating, or ready. Therefore there was no point trying to assert which state the endpoint is.

jrajahalme commented on tests/10-proxy.sh:
```diff
@@ -177,9 +176,7 @@ function policy_many_ingress {
 	],
 	"toPorts": [{
 	    "ports": [{"port": "80", "protocol": "tcp"},
-		      {"port": "8080", "protool": "tcp"},
```
I wonder how we dealt with the typo in "protool" before :-)

rlenglet replied:
We ignore unknown fields when deserializing from JSON.

aanm replied:
We are maintaining this test suite? If this test is changed why there is not changed in the corresponding ginkgo tests?

tgraf replied:
@jrajahalme mentioned that he is using this manually. The change itself is unrelated to this PR we can factor it out and separate it as a separate PR.

jrajahalme reviewed (dismissed):
Overall I think this PR should not change the endpoint state in the API, i.e., an endpoint with the init identity should remain in the waiting-for-identity state without transitioning away from that state before receiving the real identity.

Other points are minor comments you may address at your pleasure :-)

rlenglet committed:
daemon: Define reserved:init label and set it on endpoints with no labels

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
policy: Always enable policy for reserved:init endpoints in default mode

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
api: Add "init" as supported entity

Document "init" as a supported entity in the k8s CRD schema.
Bump the CRD schema version.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
identity: Allocate reserved identities for entity reserved labels

Allocate ID 4 to "health" entities (health daemon endpoints).
Allocate ID 5 to "init" entities (initializing endpoints).

Allocate such entity IDs synchronously, since that doesn't require
any kvstore interactions.

Fixes: https://github.com/cilium/cilium/issues/3895
Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
cilium-docker: Remove constraint on endpoint state after creation

After creation, endpoints can now be in different states: creating,
waiting-for-identity, not-ready, etc.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
cilium-docker: Create veth pair on endpoint creation

Create the veth pair immediately, so that the LXC and node MAC
addresses are known at endpoint creation time, like with the CNI
plugin.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
cilium-docker: Remove now-unnecessary PATCH /endpoint/{id} API call

All the information that used to be passed in the PATCH call is
now passed in the PUT call. The remaining state change triggered by
the PATCH call was unnecessary.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
endpoint: Fix state machine to support changing endpoint's labels

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
daemon: Fix identity label update APIs

Have PUT /endpoint/{id} return status code 400 in case of failure
to set labels, instead of code 404 which was not specified in the
API.

On endpoint creation, set the endpoint's identity before adding
it into the endpoint manager so it is never exposed with no
identity.

PUT /endpoint/{id} now sets orchestration labels, not custom
labels, to be consistent with PATCH /endpoint/{id}.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
test: Handle initializing endpoints with the reserved:init identity

Update all helpers to wait for the endpoints' identity to be
different from "init" / 5 in addition to being ready.

Fix the egress test to ping the host instead of 8.8.8.8.

Update DROP_ALL tests to test the reserved:init identity, since
endpoints don't get a DROP_ALL program installed anymore
unless labels are passed to the PUT /endpoint/{id} API.

Test applying a policy to the reserved:init identity.

Signed-off-by: Romain Lenglet <romain@covalent.io>

ianvernon commented on pkg/identity/allocator.go:
```diff
@@ -91,6 +91,18 @@ func InitIdentityAllocator(owner IdentityAllocatorOwner) {
+// reserved identity, i.e. if the slice contains a single reserved
+// "reserved:*" label.
+func IdentityAllocationIsLocal(lbls labels.Labels) bool {
+	// If there only one label with the "reserved" source and a well-known key,
```
If there is only

rlenglet replied:
Fixed.

ianvernon commented on pkg/identity/allocator.go:
```diff
@@ -100,6 +112,17 @@ func AllocateIdentity(lbls labels.Labels) (*Identity, bool, error) {
 		logfields.IdentityLabels: lbls.String(),
 	}).Debug("Resolving identity")
 
+	// If there only one label with the "reserved" source and a well-known key,
```
If there is only

rlenglet replied:
Fixed.

ianvernon commented on pkg/identity/cache.go:
```diff
@@ -97,6 +93,25 @@ func LookupIdentity(lbls labels.Labels) *Identity {
+// LookupReservedIdentity looks up a reserved identity by its labels and
+// returns it if found. Returns nil if not found.
+func LookupReservedIdentity(lbls labels.Labels) *Identity {
+	// If there only one label with the "reserved" source and a well-known key,
```
If there is only

rlenglet replied:
Fixed.

ianvernon commented on daemon/endpoint.go:
```diff
@@ -315,11 +315,12 @@ func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Respon
 	}
 
+	// TODO: Do something with the labels?
+	// addLabels := labels.NewLabelsFromModel(params.Endpoint.Labels)
```
Meaning of this TODO?

rlenglet replied:
The meaning is that we have been completely ignoring the labels passed to the PATCH call until now. This TODO highlights that fact.

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -2129,52 +2130,55 @@ func (e *Endpoint) getIDandLabels() string {
-				delete(newLabels.Custom, k)
-			}
+		if newLabels.Custom[k] != nil {
+			delete(newLabels.Custom, k)
```
Do we not mark custom labels as "disabled"? I guess an argument for not marking them as such is because since they were added by the user and not the orchestration system, they can be deleted without being marked as disabled. Is this correct?

rlenglet replied:
Correct.

ianvernon commented on test/helpers/cmd.go:
```diff
@@ -93,6 +94,22 @@ func (res *CmdRes) ExpectContains(data string, optionalDescription ...interface{
+// ExpectDoesntContain asserts that a string is not contained in the stdout of
+// the executed command. It accepts an optional parameter that can be used to
+// annotate failure messages.
+func (res *CmdRes) ExpectDoesntContain(data string, optionalDescription ...interface{}) bool {
```
ExpectDoesntContain --> ExpectDoesNotContain

joestringer replied:
`ExpectOutputLacks()`?

rlenglet replied:
`ExpectDoesNotContain` fits better with the test framework's naming. Renamed.

ianvernon commented on test/helpers/cmd.go:
```diff
@@ -93,6 +94,22 @@ func (res *CmdRes) ExpectContains(data string, optionalDescription ...interface{
+// ExpectDoesntMatchRegexp asserts that the stdout of the executed command
+// doesn't match the regexp. It accepts an optional parameter that can be used
+// to annotate failure messages.
+func (res *CmdRes) ExpectDoesntMatchRegexp(regexp string, optionalDescription ...interface{}) bool {
```
ExpectDoesntMatchRegexp --> ExpectDoesNotMatchRegexp

rlenglet replied:
Done.

ianvernon reviewed (commented):
Looks great overall, mainly just nits with comments / some questions.

joestringer commented on daemon/endpoint.go:
```diff
@@ -160,24 +159,26 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
+	if len(addLabels) > 0 {
+		addLabels, _, ok := checkLabels(addLabels, nil)
+		if !ok {
+			return PutEndpointIDInvalidCode, fmt.Errorf("No valid label")
```
`checkLabels()` performs this check:

```
if len(addLabels) == 0 && len(delLabels) == 0 {
    return nil, nil, false
```

This is impossible, since immediately above you are ensuring that the length of `addLabels` is > 0.

rlenglet replied:
`checkLabels` calls `FilterLabels`, which can return an empty label list and trigger that condition.

joestringer commented on pkg/controller/manager.go:
```diff
@@ -63,6 +63,7 @@ func (m *Manager) UpdateController(name string, params ControllerParams) *Contro
 
 	if oldCtrl, ok := m.controllers[name]; ok {
 		oldCtrl.stopController()
+		m.removeController(oldCtrl)
```
Is this a bugfix that should be backported?

tgraf replied:
The real bug here is that a new uuid is generated every time the controller is updated which is horribly wrong. Let's fix it properly outside of this PR.

tgraf replied:
I've filed https://github.com/cilium/cilium/issues/4066

rlenglet replied:
Merged separately.

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -1315,10 +1323,12 @@ func (epMap *EndpointMap) GetPolicyStatus() map[models.EndpointPolicyEnabled]int
 	for _, ep := range *epMap {
-		if ep.Status.State != models.EndpointStateReady {
+		// Consider an endpoint with reserved identity 5 (reserved:init) as not ready.
+		if ep.Status.State != models.EndpointStateReady || ep.Status.Identity.ID == 5 {
```
If the `5` here referred to the reserved:init identity by name, then the comment would be unnecessary.

joestringer commented on daemon/endpoint.go:
```diff
@@ -160,21 +159,27 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
+	if len(addLabels) > 0 {
+		addLabels, _, ok := checkLabels(addLabels, nil)
+		if !ok {
+			return PutEndpointIDInvalidCode, fmt.Errorf("No valid label")
```
This case can't be hit, check the implementation of `checkLabels()`.

rlenglet replied:
It can, if all the labels in `addLabels` are filtered out by the call to `FilterLabels` made in `checkLabels`.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2240,22 +2242,39 @@ func (e *Endpoint) identityResolutionIsObsolete(myChangeRev int) bool {
 
+// Must be called with e.Mutex NOT held.
 func (e *Endpoint) runLabelsResolver(owner Owner, myChangeRev int) {
+	e.Mutex.Lock()
```
Not really clean but probably the best compromise.

tgraf commented on plugins/cilium-docker/driver/driver.go:
```diff
@@ -399,16 +399,6 @@ func (driver *driver) joinEndpoint(w http.ResponseWriter, r *http.Request) {
-		State: models.EndpointStateWaitingForIdentity,
-	}
-
-	if err = driver.client.EndpointPatch(endpointPkg.NewCiliumID(old.ID), ep); err != nil {
```
This means that we now connect the container on endpoint creation before the actual join. That may be unexpected. Did you think about this?

tgraf replied:
All changes exact for the relaxation of the state requirement should be done in a separate PR. It has nothing to do with the init identity itself.

aanm replied:
Agreed, I'm not that confident with these changes, can you make this on a different PR?

tgraf replied:
Discussed during community meeting. The endpoint is not connected before the join is done. The veth pair name is still returned in JoinResponse.

tgraf commented on daemon/endpoint.go:
```diff
@@ -160,24 +159,26 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
-		}).WithError(err).Error("Could not add labels while creating an ep")
-		return code, err
-	}
+	ep.SetIdentityLabels(d, addLabels)
```
Move this up to before `endpointmanager.AddEndpoint()` so we never even expose the endpoint before it has the init identity.

rlenglet replied:
Done.

tgraf commented on pkg/controller/controller.go:
```diff
@@ -130,6 +130,7 @@ type Controller struct {
 	lastErrorStamp    time.Time
 	uuid              string
 	stop              chan struct{}
+	stopForUpdate     chan struct{}
```
Please submit this commit separately.

tgraf reviewed (dismissed):
This is an awesome PR but contains a ton of unrelated changes. In particular the last commits are mostly unclear on why certain changes are required. I assume the changes are needed to make the tests pass but there is no context provided at all on why these changes are needed.

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -561,7 +561,7 @@ var (
 				Description: "FromEntities is a list of special entities which the endpoint " +
 					"subject to the rule is allowed to receive connections from. Supported " +
-					"entities are `world`, `cluster` and `host`",
+					"entities are `world`, `cluster`, `host`, and `init`",
```
schema validation change needs version bumping on line 44

aanm reviewed (dismissed):
Can some of the commits be made on a separate PR? My main concern is around the cilium-docker changes

rlenglet committed:
k8s: Don't add namespace labels into reserved:init endpoint selectors

Signed-off-by: Romain Lenglet <romain@covalent.io>

tgraf commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -112,7 +112,7 @@ func parseToCiliumIngressRule(namespace string, inRule, retRule *api.Rule) {
 					// FromEndpoints selector. If omitted, we limit the
 					// scope to the namespace the policy lives in.
-					if _, ok := retRule.Ingress[i].FromRequires[j].MatchLabels[podPrefixLbl]; !ok {
+					if !retRule.Ingress[i].FromRequires[j].HasKey(podPrefixLbl) {
```
This looks like a bug so we should submit it separately so we can backport it

aanm replied:
The same goes for line 178

```golang
if _, ok := retRule.Egress[i].ToRequires[j].MatchLabels[podPrefixLbl]; !ok {
	retRule.Egress[i].ToRequires[j].MatchLabels[podPrefixLbl] = namespace
}
```

rlenglet committed:
endpointmanager: Don't generate new endpoints waiting-for-identity

The initial generation for a new endpoint is now handled by setting
the initial identity.

Signed-off-by: Romain Lenglet <romain@covalent.io>

tgraf reviewed (approved):
LGTM, can be merged as soon as #4146 is merged
